### PR TITLE
Add the batch_norm op

### PR DESCRIPTION
**Pull request by haifeng-jin**
The main reason for adding this op is to resolve the torch memory issue.
This allows torch to use the fused op for batch norm, which saves memory.

# Benchmarking

Measured by averaging 100 steps. 

| backend | batch size | fit (ms/step) | predict (ms/step) | link |
| -------- | ------- | ------- | ------- | ------- |
| JAX  (before) | 128    | 459 | 178 | [link](https://github.com/haifeng-jin/keras-benchmarking/blob/main/batch_norm_op_jax_before.ipynb) |
| JAX  (after) | 128   | 456 | 171 | [link](https://github.com/haifeng-jin/keras-benchmarking/blob/main/batch_norm_op_jax_after.ipynb) |
| Torch (before) | 64 | 397 | 141 | [link](https://github.com/haifeng-jin/keras-benchmarking/blob/main/batch_norm_op_torch_before.ipynb) |
| Torch (after) | 64  | 321 | 126 | [link](https://github.com/haifeng-jin/keras-benchmarking/blob/main/batch_norm_op_torch_after.ipynb) |
| Torch (before) | 98   | OOM |  | [link](https://github.com/haifeng-jin/keras-benchmarking/blob/main/batch_norm_op_oom_torch_before.ipynb) |
| Torch (after) | 98    | Passed |  | [link](https://github.com/haifeng-jin/keras-benchmarking/blob/main/batch_norm_op_oom_torch_after.ipynb) |

### CODE REVIEW

File: keras/backend/jax/nn.py [hunk 1]
```diff
@@ -527,3 +527,22 @@ def moments(x, axes, keepdims=False, synchronized=False):
         mean = cast(mean, ori_dtype)
         variance = cast(variance, ori_dtype)
     return mean, variance
+
+
+def batch_norm(x, mean, variance, axis, offset=None, scale=None, epsilon=1e-3):
```
Comment (fchollet) [inline keras/backend/jax/nn.py:532]: Rename to `batch_normalization`

File: keras/backend/torch/nn.py [hunk 2]
```diff
@@ -686,3 +686,42 @@ def moments(x, axes, keepdims=False, synchronized=False):
...
+    if axis == 1:
+        return _batch_norm()
+
+    if axis < 0:
+        axis = len(x.shape) + axis
+
+    order = list(range(len(x.shape)))
+    order.pop(axis)
+    order.insert(1, axis)
+    x = x.permute(order)
```
Comment (fchollet) [inline keras/backend/torch/nn.py:720]: It sounds to me that permute + native BN op + permute will be more expensive memory wise than doing a low level BN implementation like before. Permutation is expensive.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

**Comment by fchollet** [comment]
Note that we should validate the final code via a benchmark on a convnet (with all backends). I'm not worried about the TF side, but the JAX and torch side need some numbers.

File: keras/backend/torch/nn.py (hunk 2 above)
Comment (haifeng-jin) [reply keras/backend/torch/nn.py:720]: According to the [torch docs](https://pytorch.org/docs/stable/generated/torch.permute.html), the permute is a view of the original tensor. I guess it is not that expensive.

Comment (haifeng-jin) [reply keras/backend/torch/nn.py:720]: If set to use channels_first, the permute would not run as the function returns before reaching it.

**Comment by haifeng-jin** [comment]
> Note that we should validate the final code via a benchmark on a convnet (with all backends). I'm not worried about the TF side, but the JAX and torch side need some numbers.

Sure. I can benchmark different backends with a convnet using a commit before this change and a commit after this change to compare the numbers.

File: keras/backend/jax/nn.py (hunk 1 above)
Comment (haifeng-jin) [reply keras/backend/jax/nn.py]: Done.

**Comment by haifeng-jin** [comment]
Numbers updated in the description.
We have speedups on torch and jax.
Memory saved on torch.

File: keras/ops/nn.py [hunk 3]
```diff
@@ -1637,3 +1637,87 @@ def moments(x, axes, keepdims=False, synchronized=False):
...
+        mean: A mean vector of the same length as the `axis` dimension of the
+            input thensor.
+        variance: A variance vector of the same length as the `axis` dimension
+            of the input tensor.
+        axis: Integer, the axis that should be normalized.
+        offset: An offset vector of the same length as the `axis` dimension of
+            the input tensor. If not None, `offset` is added to the normalized
+            tensor. Defaults to None.
+        scale: A scale vector of the same length as the `axis` dimension of the
+            input tensor. If not None, the normalized tensor is multiplied by
```
Comment (fchollet) [inline keras/ops/nn.py:1692]: Use backticks around `None` everywhere

**Review by fchollet — COMMENTED** [review]
Code LGTM. Thank you!

File: keras/ops/nn.py (hunk 3 above)
Comment (haifeng-jin) [reply keras/ops/nn.py]: Done.
